tracer/src/Datadog.Trace/AppSec/Waf/ConfigurationState.cs

```diff
@@ -200,99 +270,128 @@ public void ApplyStoredFiles()
                     }
                 }
+            }
 
-                // only deserialize and apply asm_features as it will decide if asm gets toggled on and if we deserialize all the others
-                // (the enable of auto user instrumentation as added to asm_features)
-                _asmFeatureProduct.ProcessUpdates(this, asmFeaturesToUpdate);
-                _asmFeatureProduct.ProcessRemovals(this, asmFeaturesToRemove);
+            ApplyAsmFeatures(AppsecEnabled);
+            IncomingUpdateState.ShouldUpdateAppsec = !IncomingUpdateState.ShouldInitAppsec && AppsecEnabled && hasUpdateConfigurations;
 
-                EnableAsm = !AsmFeaturesByFile.IsEmpty() && AsmFeaturesByFile.All(a => a.Value?.Enabled is null or true);
+            if (IncomingUpdateState.ShouldUpdateAppsec || IncomingUpdateState.ShouldInitAppsec)
+            {
+                ApplyStoredFiles();
+            }
+        }
+    }
 
-                // empty, one value, or all values the same are valid states, anything else is an error
-                var autoUserInstrumMode = AutoUserInstrumByFile.Values.FirstOrDefault(x => x?.Mode is not null);
-                if (autoUserInstrumMode == null ||
-                    AutoUserInstrumByFile.All(x => x.Value?.Mode is null || x.Value?.Mode == autoUserInstrumMode?.Mode))
-                {
-                    AutoUserInstrumMode = autoUserInstrumMode?.Mode?.ToLowerInvariant();
-                }
-                else
-                {
-                    AutoUserInstrumMode = "unknown value";
-                    Log.Error(
-                        "AutoUserInstrumMode was 'unknown value', source data: {AutoUserInstrumByFile}",
-                        string.Join(",", AutoUserInstrumByFile.Values.Select(x => x?.Mode)));
-                }
+    private void ApplyAsmFeatures(bool appsecCurrentlyEnabled)
+    {
+        var change = false;
+        // only deserialize and apply asm_features as it will decide if asm gets toggled on and if we deserialize all the others
+        // (the enable of auto user instrumentation as added to asm_features)
+        if (_fileUpdates.TryGetValue(RcmProducts.AsmFeatures, out var updates))
+        {
+            _asmFeatureProduct.ProcessUpdates(this, updates);
+            change = true;
+        }
+
+        if (_fileRemoves.TryGetValue(RcmProducts.AsmFeatures, out var removals))
+        {
+            _asmFeatureProduct.ProcessRemovals(this, removals);
+            change = true;
+        }
+
+        if (!change) { return; }
 
-                AutoUserInstrumMode = autoUserInstrumMode?.Mode?.ToLowerInvariant();
+        // normally CanBeToggled should not need a check as asm_features capacity is only sent if AppSec env var is null, but still guards it in case
+        if (_canBeToggled)
+        {
+            var rcmEnable = !AsmFeaturesByFile.IsEmpty() && AsmFeaturesByFile.All(a => a.Value?.Enabled is null or true);
```

Comment:
```suggestion
            var rcmEnable = !AsmFeaturesByFile.Any(a => a.Value?.Enabled is false);
```

Reply:
thanks, I ll change with Any but I'll still keep ` !AsmFeaturesByFile.IsEmpty()` as otherwise if files are removed and appsec was on, it's not gonna turn it back off anymoee